### PR TITLE
[base-ubuntu] - remove existing ubuntu user

### PR DESCRIPTION
**Pull request by gauravsaini04**
**Devcontainer Names**

* base-ubuntu

**Description**

* This PR attemtps to make the following changes:
	- Removes the existing `ubuntu` user provided by the base-image for `ubuntu-noble` base image i.e. existing user blocking uid & gid 1000
	
**_Changelog_** 

* Changed the `Dockerfile` and wrote a test case for the devcontainer image

**Checklist**

* [x] Checked that applied changes work as expected

### CODE REVIEW

File: src/base-ubuntu/.devcontainer/Dockerfile [hunk 1]
```diff
@@ -3,6 +3,33 @@ FROM buildpack-deps:${VARIANT}-curl
...
+        else \
+            # Check if vscode user exists, if not create it
+            if ! id "vscode" &>/dev/null; then \
+                echo "Creating user vscode" && useradd -m vscode || echo "Failed to create vscode user"; \
+            fi; \
+            # Modify vscode user and group only if they exist
+            echo "Modifying user vscode UID" && usermod -u 1000 vscode || echo "Failed to modify vscode UID"; \
+            echo "Modifying group vscode GID" && groupmod -g 1000 vscode || echo "Failed to modify vscode GID"; \
+        fi; \
+    else \
```
Comment (samruddhikhandale) [inline src/base-ubuntu/.devcontainer/Dockerfile:29]: The `vscode` user is created by `common-utils` Feature. Hence, you don't need to create it in the dockerfile. 

I was referring to re-adding https://github.com/devcontainers/images/pull/1036/files#diff-eaa1990d44094d26a8143b63dc994e880f6b4fa69140f3b6bf270c4acecbf22aL10-L11

File: src/base-ubuntu/.devcontainer/Dockerfile [hunk 2]
```diff
@@ -3,6 +3,33 @@ FROM buildpack-deps:${VARIANT}-curl
...
 LABEL dev.containers.features="common"
 
+ARG VARIANT
+ENV VARIANT=${VARIANT}
+
+RUN if [ "$VARIANT" = "noble" ]; then \
+        echo "Variant contains noble"; \
+        # Check if ubuntu user exists and delete it if it does
+        if id "ubuntu" &>/dev/null; then \
+            echo "Deleting user ubuntu" && userdel -f -r ubuntu || echo "Failed to delete ubuntu user"; \
```
Comment (samruddhikhandale) [inline src/base-ubuntu/.devcontainer/Dockerfile:13]: Along with deleting the user, can you also remove user's home directory and delete the group associated with the user? 

This should free up the 1000 uid and gid

File: src/base-ubuntu/.devcontainer/Dockerfile [hunk 3]
```diff
@@ -3,6 +3,33 @@ FROM buildpack-deps:${VARIANT}-curl
...
+        else \
+            echo "User ubuntu does not exist"; \
+        fi; \
+        # Double-check if UID 1000 is still in use
+        if id -u 1000 &>/dev/null; then \
+            echo "UID 1000 is still in use. Exiting."; \
+        else \
```
Comment (samruddhikhandale) [inline src/base-ubuntu/.devcontainer/Dockerfile:20]: This is fine for debugging, but when you open a final PR, let's remove this.
Let our tests validate this.

**Review by samruddhikhandale — CHANGES_REQUESTED** [review]

File: src/base-ubuntu/.devcontainer/Dockerfile (hunk 1 above)
Comment (gauravsaini04) [reply src/base-ubuntu/.devcontainer/Dockerfile]: added back again 👍

File: src/base-ubuntu/.devcontainer/Dockerfile (hunk 2 above)
Comment (gauravsaini04) [reply src/base-ubuntu/.devcontainer/Dockerfile:13]: userdel command supposed to delete the user's home dir as well, have added command to delete user's group

File: src/base-ubuntu/.devcontainer/Dockerfile (hunk 3 above)
Comment (gauravsaini04) [reply src/base-ubuntu/.devcontainer/Dockerfile]: ok

**Comment by gauravsaini04** [comment]
```
https://github.com/devcontainers/images/actions/runs/10534973968/job/29193352531?pr=1170#step:3:2854
```
The failing tests would pass in the dev codespace env.

File: src/base-ubuntu/.devcontainer/Dockerfile [hunk 4]
```diff
@@ -3,6 +3,22 @@ FROM buildpack-deps:${VARIANT}-curl
...
+        # Check if ubuntu user exists and delete it if it does
+        if id "ubuntu" &>/dev/null; then \
+            echo "Deleting user ubuntu" && userdel -f -r ubuntu || echo "Failed to delete ubuntu user"; \
+            echo "Deleting group associated with ubuntu" && groupdel ubuntu || echo "Failed to delete group associated with ubuntu user"; \
+        else \
+            echo "User ubuntu does not exist"; \
+        fi; \
+    else \
+        echo "Variant does not contain noble"; \
+    fi > /tmp/logfile.txt 2>&1
```
Comment (samruddhikhandale) [inline src/base-ubuntu/.devcontainer/Dockerfile:20]:
```suggestion
    fi
```

Comment (samruddhikhandale) [reply src/base-ubuntu/.devcontainer/Dockerfile:20]: It's good for debugging, but let's not have it in the image

File: src/dotnet/.devcontainer/Dockerfile [hunk 5]
```diff
@@ -1,7 +1,23 @@
-ARG VARIANT=8.0-bookworm-slim
+ARG VARIANT=8.0-noble
```
Comment (samruddhikhandale) [inline src/dotnet/.devcontainer/Dockerfile:1]: Revert this change and add a test to validate noble.

File: src/base-ubuntu/.devcontainer/Dockerfile [hunk 6]
```diff
@@ -3,6 +3,22 @@ FROM buildpack-deps:${VARIANT}-curl
...
 LABEL dev.containers.features="common"
 
+ARG VARIANT
+ENV VARIANT=${VARIANT}
+
+RUN if [ "$VARIANT" = "noble" ]; then \
+        echo "Variant contains noble"; \
+        # Check if ubuntu user exists and delete it if it does
+        if id "ubuntu" &>/dev/null; then \
+            echo "Deleting user ubuntu" && userdel -f -r ubuntu || echo "Failed to delete ubuntu user"; \
```
Comment (samruddhikhandale) [inline src/base-ubuntu/.devcontainer/Dockerfile:13]: For debugging only: Can you verify that the UID and GID are no longer in use with `getent passwd 1000 && getent group 1000` ?

File: src/base-ubuntu/.devcontainer/Dockerfile (hunk 2 above)
Comment (samruddhikhandale) [reply src/base-ubuntu/.devcontainer/Dockerfile:13]: Can you check by removing entries from `/etc/passwd` and `/etc/group`

File: src/base-ubuntu/.devcontainer/Dockerfile (hunk 6 above)
Comment (gauravsaini04) [reply src/base-ubuntu/.devcontainer/Dockerfile:13]: yes, they are no longer in use when checked with getent

File: src/base-ubuntu/.devcontainer/Dockerfile (hunk 2 above)
Comment (gauravsaini04) [reply src/base-ubuntu/.devcontainer/Dockerfile:13]: no entries were found for 1000 uid and gid in the files after removal

**Review by samruddhikhandale — CHANGES_REQUESTED** [review]

File: src/base-ubuntu/.devcontainer/Dockerfile [hunk 7]
```diff
@@ -3,6 +3,25 @@ FROM buildpack-deps:${VARIANT}-curl
 
 LABEL dev.containers.features="common"
 
+ARG VARIANT
+ENV VARIANT=${VARIANT}
+
+RUN if [ "$VARIANT" = "noble" ]; then \
+        echo "Variant contains noble"; \
+        # Check if ubuntu user exists and delete it if it does
+        if id "ubuntu" &>/dev/null; then \
```
[the remark this answers is not in the document]
Comment (gauravsaini04) [reply src/base-ubuntu/.devcontainer/Dockerfile:12]: No actually `userdel` command already deleted everything related to `Ubuntu` user occupying `1000` uid and gid so it's just a failed check after trying to delete `ubuntu` group later.

File: src/base-ubuntu/test-project/test.sh [hunk 8]
```diff
@@ -19,7 +19,7 @@ check "gitconfig-file-location" sh -c "ls /etc/gitconfig"
 check "gitconfig-contains-name" sh -c "cat /etc/gitconfig | grep 'name = devcontainers'"
 check "usr-local-etc-config-does-not-exist" test ! -f "/usr/local/etc/gitconfig"
 
-check "uid" bash -c "id -u vscode | grep 1001" 
+check "uid" bash -c "id -u vscode | grep 1001"
```
[the remark this answers is not in the document]
Comment (gauravsaini04) [reply src/base-ubuntu/test-project/test.sh:22]: ok

[the remark this answers is not in the document]
Comment (gauravsaini04) [reply src/base-ubuntu/test-project/test.sh:22]: sure

File: src/base-ubuntu/.devcontainer/Dockerfile (hunk 7 above)
[the remark this answers is not in the document]
Comment (samruddhikhandale) [reply src/base-ubuntu/.devcontainer/Dockerfile:12]: If that's the case, then should we remove https://github.com/devcontainers/images/pull/1170/files#diff-8bc3ab047cce2f0dc035c0856710e0988c4a9bb82687cfa3d182866665ebad42R14

[the remark this answers is not in the document]
Comment (samruddhikhandale) [reply src/base-ubuntu/.devcontainer/Dockerfile:12]: @gauravsaini04 Let's only keep the commands which are actually needed for deleting `ubuntu` user and clearing up 1000 uids/gids

[the remark this answers is not in the document]
Comment (gauravsaini04) [reply src/base-ubuntu/.devcontainer/Dockerfile:12]: sure thanks for the inputs..

[the remark this answers is not in the document]
Comment (samruddhikhandale) [reply src/base-ubuntu/.devcontainer/Dockerfile:12]: > @gauravsaini04 Let's only keep the commands which are actually needed for deleting ubuntu user and clearing up 1000 uids/gids

@gauravsaini04 I don't think the code changed post my last review. Am I missing anything?

[the remark this answers is not in the document]
Comment (gauravsaini04) [reply src/base-ubuntu/.devcontainer/Dockerfile:10]: made more changes

File: src/base-ubuntu/.devcontainer/Dockerfile [hunk 9]
```diff
@@ -3,6 +3,17 @@ FROM buildpack-deps:${VARIANT}-curl
 
 LABEL dev.containers.features="common"
 
+ARG VARIANT
+ENV VARIANT=${VARIANT}
```
Comment (samruddhikhandale) [inline src/base-ubuntu/.devcontainer/Dockerfile:7]:
```suggestion
ARG VARIANT
```

Do we need `VARIANT` at runtime? I wonder if having it at build time should be sufficient in this case?

File: src/base-ubuntu/.devcontainer/Dockerfile [hunk 10]
```diff
@@ -3,6 +3,17 @@ FROM buildpack-deps:${VARIANT}-curl
...
+
+RUN if [ "$VARIANT" = "noble" ]; then \
+        if id "ubuntu" &>/dev/null; then \
+            echo "Deleting user ubuntu" && userdel -f -r ubuntu || echo "❕Failed to delete ubuntu user"; \
+        else \
+            echo "✔️ User ubuntu does not exist"; \
```
Comment (samruddhikhandale) [inline src/base-ubuntu/.devcontainer/Dockerfile:13]:
```suggestion
            echo "User 'ubuntu' does not exist for $VARIANT"; \
```

File: src/base-ubuntu/.devcontainer/Dockerfile [hunk 11]
```diff
@@ -3,6 +3,17 @@ FROM buildpack-deps:${VARIANT}-curl
 
 LABEL dev.containers.features="common"
 
+ARG VARIANT
+ENV VARIANT=${VARIANT}
+
+RUN if [ "$VARIANT" = "noble" ]; then \
+        if id "ubuntu" &>/dev/null; then \
+            echo "Deleting user ubuntu" && userdel -f -r ubuntu || echo "❕Failed to delete ubuntu user"; \
```
Comment (samruddhikhandale) [inline src/base-ubuntu/.devcontainer/Dockerfile:11]:
```suggestion
            echo "Deleting user 'ubuntu'  for $VARIANT" && userdel -f -r ubuntu || echo "Failed to delete ubuntu user for $VARIANT"; \
```

**Review by samruddhikhandale — CHANGES_REQUESTED** [review]

File: src/base-ubuntu/.devcontainer/Dockerfile (hunk 9 above)
Comment (gauravsaini04) [reply src/base-ubuntu/.devcontainer/Dockerfile]: yes that's right, works without env variable declaration as well..

File: src/base-ubuntu/.devcontainer/Dockerfile (hunk 10 above)
Comment (gauravsaini04) [reply src/base-ubuntu/.devcontainer/Dockerfile]: done

File: src/base-ubuntu/.devcontainer/Dockerfile (hunk 11 above)
Comment (gauravsaini04) [reply src/base-ubuntu/.devcontainer/Dockerfile]: done

**Review by samruddhikhandale — APPROVED** [review]

**Pull request merged**